### PR TITLE
ENH Optimized Preprocessing Algorithm nips 2016

### LINKED ISSUE #1028: ENH Add mitigation algorithm from "Optimized Pre-Processing for Discrimination Prevention" by Calmon et al.
The paper [Optimized Pre-Processing for Discrimination Prevention](https://papers.nips.cc/paper/2017/file/9a49a25d845a483fae4be7e341368e36-Paper.pdf) by Calmon et al. introduces a preprocessing algorithm for demographic parity while limiting the number of individual distortions. The goal of this task is to add the algorithm described in this paper to Fairlearn.

* The original implementation by the authors of the paper can be found [here](https://github.com/fair-preprocessing/nips2017).
* An existing implementation can be found in AIF360 [`OptimPreproc`](https://aif360.readthedocs.io/en/stable/modules/generated/aif360.algorithms.preprocessing.OptimPreproc.html).

Completing this item requires:
- [ ] technique code in `fairlearn.preprocessing`
- [ ] unit tests in `test.unit.preprocessing`
- [ ] descriptive API reference (directly in the docstring)
- [ ] a short user guide in `docs.user_guide.mitigation.rst`

A fully fledged example notebook is not required.

To claim this task please respond below. Of course, you can also ask questions!

### PR DESCRIPTION
**Pull request by rakesh9177**
## Description

closes #1028 

## Tests

- [ ] no new tests required
- [x] new tests added
- [ ] existing tests adjusted

## Documentation

- [ ] no documentation changes needed
- [ ] user guide added or updated
- [ ] API docs added or updated
- [ ] example notebook added or updated

### CODE REVIEW

File: fairlearn/preprocessing/_optimPreproc.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+            dfP_noY = dfP_noY.divide(dfP_noY.sum(axis=1), axis=0)
+
+            df_transformed = _apply_randomized_mapping(
+                df,
+                dfP_noY,
+                features=D_features + X_features,
+                random_seed=self.seed,
+            )
+        return df_transformed
+        """
```
Comment (romanlutz) [inline fairlearn/preprocessing/_optimPreproc.py:105]: Can this commented out code be removed?

File: fairlearn/preprocessing/_optimPreproc_helper.py [hunk 2]
```diff
@@ -0,0 +1,393 @@
+import numpy as np
```
Comment (romanlutz) [inline fairlearn/preprocessing/_optimPreproc_helper.py:1]: Copyright note missing at the top of every file. You probably need ot run flake8 and fix all the issues.

File: fairlearn/preprocessing/_optimPreproc_helper.py [hunk 3]
```diff
@@ -0,0 +1,393 @@
...
+            )
+
+        # epsilon definition for discrimination control
+        # epsilon = .1
+        # disc_upper = Variable()
+        # disc_lower = Variable()
+
+        # Discrimination control
+        for d in range(self.dfMask_Pxyd_to_Pd.shape[1]):
+            # constraints.append(sum(kl_div(PYhgD[d,:].T,PyMarginal))<=self.epsilon)
```
Comment (romanlutz) [inline fairlearn/preprocessing/_optimPreproc_helper.py:319]: Can this commented out code be removed?

File: requirements.txt [hunk 4]
```diff
@@ -6,3 +6,5 @@ pandas>=2.0.3
 pyarrow>=15
 scikit-learn>=1.2.1
 scipy>=1.9.3
+cvxpy
```
Comment (romanlutz) [inline requirements.txt:9]: This may be a problem... I believe this has come up in the past, too. I wouldn't mind it as a soft dependency. Paging @fairlearn/fairlearn-maintainers 🙂

File: test/unit/preprocessing/Discrimation_remover_optimized_preprocess/test_optimPreproc.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
+import pandas as pd
+import numpy as np
+
+from fairlearn.datasets import fetch_adult
+from fairlearn.preprocessing import OptimizedPreprocessor
+from fairlearn.preprocessing._optimPreproc_helper import DTools
+
+# import warnings
```
Comment (romanlutz) [inline test/unit/preprocessing/Discrimation_remover_optimized_preprocess/test_optimPreproc.py:8]: Can this commented out code be removed?

File: test/unit/preprocessing/Discrimation_remover_optimized_preprocess/test_optimPreproc.py [hunk 6]
```diff
@@ -0,0 +1,144 @@
+import pandas as pd
```
Comment (romanlutz) [inline test/unit/preprocessing/Discrimation_remover_optimized_preprocess/test_optimPreproc.py:1]: nit: folder name has a typo (discrimination) and should start with lower case, file name should be snake case

File: fairlearn/preprocessing/_optimPreproc.py [hunk 7]
```diff
@@ -0,0 +1,213 @@
+import numpy as np
+import pandas as pd
+from ._optimPreproc_helper import DTools
+
+from sklearn.base import BaseEstimator, TransformerMixin
+
+
+class OptimizedPreprocessor:
```
[the remark this answers is not in the document]
Comment (hildeweerts) [reply fairlearn/preprocessing/_optimPreproc.py:8]: I would not go for something like "discrimination-reducing" as that could imply the pre-processing algorithm maps neatly upon legal notions of discrimination.

The technique aims to identify a transformation of the data that minimizes the difference between the original distribution of X and Y and the distribution of the pre-processed data ("utility"), subject to a demographic parity ratio constraint ("discrimination control") and an expected individual "distortion" constraint ("distortion control"). The last component determines which transformations are penalized (e.g. swapping labels, changing particular features, etc.). So perhaps a name related to demographic parity and individual distortions would make sense?

[the remark this answers is not in the document]
Comment (rakesh9177) [reply fairlearn/preprocessing/_optimPreproc.py:8]: Added description

File: fairlearn/preprocessing/_optimPreproc.py [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+
+        self.opt.optimize(
+            epsilon=self.optim_options["epsilon"],
+            dlist=self.optim_options["dlist"],
+            verbose=self.verbose,
+        )
+
+        self.opt.computeMarginals()
+
+        """
```
[the remark this answers is not in the document]
Comment (rakesh9177) [reply fairlearn/preprocessing/_optimPreproc.py]: Removed the comments

File: fairlearn/preprocessing/_optimPreproc.py (hunk 1 above)
Comment (rakesh9177) [reply fairlearn/preprocessing/_optimPreproc.py]: Removed the comments

File: fairlearn/preprocessing/_optimPreproc_helper.py [hunk 9]
```diff
@@ -0,0 +1,393 @@
+import numpy as np
+import pandas as pd
+from cvxpy import *
+
+
+class DTools:
```
[the remark this answers is not in the document]
Comment (rakesh9177) [reply fairlearn/preprocessing/_optimPreproc_helper.py:6]: Used same name as used by author of paper, but we can brainstorm the name for this one. Thank you!

File: fairlearn/preprocessing/_optimPreproc_helper.py [hunk 10]
```diff
@@ -0,0 +1,393 @@
...
+        # compute mask that reduces Pxyd to Pd
+        self.dfMask_Pxyd_to_Pd = pd.DataFrame(
+            np.zeros((len(self.dfP), self.dfD_to_Y_address.shape[1])),
+            index=self.dfP.index,
+            columns=self.dfD_to_Y_address.columns,
+        )
+        self.dfMask_Pxyd_to_Pd = self.getMask(self.dfMask_Pxyd_to_Pd)
+
+    # method for passing the distortion value
+    def setDistortion(self, Dclass, clist=[]):
```
[the remark this answers is not in the document]
Comment (rakesh9177) [reply fairlearn/preprocessing/_optimPreproc_helper.py]: Done!, thanks!

File: fairlearn/preprocessing/_optimPreproc_helper.py (hunk 3 above)
Comment (rakesh9177) [reply fairlearn/preprocessing/_optimPreproc_helper.py]: Removed the comment, thank you!

File: test/unit/preprocessing/Discrimation_remover_optimized_preprocess/test_optimPreproc.py (hunk 5 above)
Comment (rakesh9177) [reply test/unit/preprocessing/Discrimation_remover_optimized_preprocess/test_optimPreproc.py]: comment removed, thank you!

File: test/unit/preprocessing/Discrimation_remover_optimized_preprocess/test_optimPreproc.py (hunk 6 above)
Comment (rakesh9177) [reply test/unit/preprocessing/Discrimation_remover_optimized_preprocess/test_optimPreproc.py:1]: changed to lower case, thank you!

File: fairlearn/preprocessing/_optimPreproc.py [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+
+from sklearn.base import BaseEstimator, TransformerMixin
+
+
+class OptimizedPreprocessor:
+    def __init__(self, optimizer, optim_options, verbose=False, seed=None) -> None:
+        super().__init__()
+
+        self.seed = seed
+        self.optimizer = optimizer
```
[the remark this answers is not in the document]
Comment (rakesh9177) [reply fairlearn/preprocessing/_optimPreproc.py:27]: You are correct! Thank you! I removed it

File: fairlearn/preprocessing/_optimPreproc.py [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+        )
+        self.opt.optimize(
+            epsilon=self.optim_options["epsilon"],
+            dlist=self.optim_options["dlist"],
+            verbose=self.verbose,
+        )
+        self.opt.compute_marginals()
+        """
+
+    def transform(self, df, X_features, Y_features, D_features, transform_Y=False):
```
[the remark this answers is not in the document]
Comment (rakesh9177) [reply fairlearn/preprocessing/_optimPreproc.py:55]: When features are being transformed, there is an option if labels(Y_features) needs to transformed along with data. If a user wishes to transform labels as well along with X features, user can make transform_Y=True.

File: fairlearn/preprocessing/_optimPreproc.py [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+class OptimizedPreprocessor:
+    def __init__(self, optimizer, optim_options, verbose=False, seed=None) -> None:
+        super().__init__()
+
+        self.seed = seed
+        self.optimizer = optimizer
+        self.optim_options = optim_options
+        self.verbose = verbose
+
+    def fit(self, df, X_features, Y_features, D_features):
```
[the remark this answers is not in the document]
Comment (rakesh9177) [reply fairlearn/preprocessing/_optimPreproc.py:31]: I can use a similar approach and it should not take much time to modify.

There are two ways Correlation remover is working right now-
1) Using numpy arrays, since this does not have feature names, index has to be passed
2)When a pandas dataframe is passed, it can take advantage of column names

Both are being passed using `sensitive_feature_ids`

File: fairlearn/preprocessing/_optimPreproc.py [hunk 14]
```diff
@@ -0,0 +1,213 @@
+import numpy as np
+import pandas as pd
+from ._optimPreproc_helper import DTools
+
+from sklearn.base import BaseEstimator, TransformerMixin
+
+
+class OptimizedPreprocessor:
+    def __init__(self, optimizer, optim_options, verbose=False, seed=None) -> None:
```
[the remark this answers is not in the document]
Comment (rakesh9177) [reply fairlearn/preprocessing/_optimPreproc.py]: Changed to arguments, and added default values as per paper for epsilon, clist and dlist, distortion function has to be passed

**Review by adrinjalali — COMMENTED** [review]
I started reviewing the API, and I see it's far from a scikit-learn compatible API. We should fix that first before merging.

Also, please apply `black` to your new files.

File: fairlearn/preprocessing/_optimPreproc.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (rakesh9177) [reply fairlearn/preprocessing/_optimPreproc.py]: Apologies for the delayed response since I had some interviews! 

I changed the code according to scikit documentation - https://scikitlearn.org/stable/modules/generated/sklearn.base.TransformerMixin.html

Please review @adrinjalali

File: fairlearn/preprocessing/_optimPreproc.py [hunk 15]
```diff
@@ -0,0 +1,134 @@
+import numpy as np
+import pandas as pd
+from ._optimPreproc_helper import DTools
+
+from sklearn.base import BaseEstimator, TransformerMixin
+
+
+class OptimizedPreprocessor:
```
[the remark this answers is not in the document]
Comment (rakesh9177) [reply fairlearn/preprocessing/_optimPreproc.py]: Adjusted the code according to -
https://scikit-learn.org/stable/modules/generated/sklearn.base.TransformerMixin.html

**Comment by TamaraAtanasoska** [comment]
hi @rakesh9177, sorry that the last reviews kind of trailed off and there was no resolution. are you still interested in finishing this PR?

**Comment by TamaraAtanasoska** [comment]
@taharallouche unless we hear from the OP, we can maybe think of working through this one as well in collaboration. I can go over the code, solve the merge conflicts, check sklearn compatibility of the code, and hand a passing CI to you to check and improve the method's workings(correctness, improve the data processing parts, performance etc.) and add missing docs if you think they aren't rightfully representing the paper descriptions.
I can start working on it next week and ping you when ready for handover. Would that be ok?

File: requirements.txt (hunk 4 above)
Comment (TamaraAtanasoska) [reply requirements.txt:5]: another ping here to the @fairlearn/fairlearn-maintainers: I have started to look into and shape up this PR (I will share the load with @taharallouche see https://github.com/fairlearn/fairlearn/pull/1340#issuecomment-2602215599). Should I spend time looking into removing this dependency for the calculations it is used for, or is it ok to add it? I am not sure if it will be useful for any other similar problems or existing refactoring. I think this is a crucial part of estimating the effort that goes into it for me, so it would be good to know. Maybe voting with +1 on this comment would do? I know it isn't so simple though (my longest lasting job was a job where no dependencies were allowed, I get it :D). 

I've never had to use the library for any projects before, maybe scientific programmers like @taharallouche have a perspective?

**Comment by taharallouche** [comment]
> Should I spend time looking into removing this dependency for the calculations it is used for, or is it ok to add it? I am not sure if it will be useful for any other similar problems or existing refactoring. I think this is a crucial part of estimating the effort that goes into it for me, so it would be good to know.

Hi @TamaraAtanasoska 

On another issue, it was suggested to add it as an optional dependency https://github.com/fairlearn/fairlearn/issues/1246#issuecomment-1604970300

I think the suggestion make sense given that it will only be required for one mitigation algorithm.

The library looks very well maintained on GitHub too.

When I'll be looking into the method/code in details, I'll see if it can be (easily) replaced with, say, scipy.optimize . I can't tell for now.

**Comment by TamaraAtanasoska** [comment]
@taharallouche yes same I can't tell at the moment just by looking at it. I am not sure if we need a consensus for this. I can maybe clean it up by leaving it there and we think about it in the later part of the process?

**Comment by TamaraAtanasoska** [comment]
Tasklist for myself:

- [x] merge current main;
- [x] initial green CI;
   - note: a specific pandas Performace warning currently ignored, some issues online say there are situations where it is unavoidable, TODO to research later;
- [ ] sklearn API compatibility **[in progress]**
- [ ] Refactor: naming(variables, functions, classes), programming style, reducing pandas operations(might need to pass this on to someone with more scipy knowledge)
